src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -223,5 +223,476 @@ function (indicator_max::IndicatorMax)(u::AbstractArray{<:Any,4},
+      # Get neighboring element ids
+      left  = cache.interfaces.neighbor_ids[1, interface]
+      right = cache.interfaces.neighbor_ids[2, interface]
+
+      # Apply smoothing
+      alpha[left]  = max(alpha_tmp[left],  0.5 * alpha_tmp[right], alpha[left])
+      alpha[right] = max(alpha_tmp[right], 0.5 * alpha_tmp[left],  alpha[right])
+    end
+
+    # Loop over L2 mortars
+    for mortar in eachmortar(dg, cache)
+      # Get neighboring element ids
+      lower = cache.mortars.neighbor_ids[1, mortar]
+      upper = cache.mortars.neighbor_ids[2, mortar]
+      large = cache.mortars.neighbor_ids[3, mortar]
+
+      # Apply smoothing
+      alpha[lower] = max(alpha_tmp[lower], 0.5 * alpha_tmp[large], alpha[lower])
+      alpha[upper] = max(alpha_tmp[upper], 0.5 * alpha_tmp[large], alpha[upper])
+      alpha[large] = max(alpha_tmp[large], 0.5 * alpha_tmp[lower], alpha[large])
+      alpha[large] = max(alpha_tmp[large], 0.5 * alpha_tmp[upper], alpha[large])
+    end
+  end
+
+  return alpha
+end
+
+
+# this method is used when the indicator is constructed as for shock-capturing volume integrals
+function create_cache(::Type{IndicatorNNRH}, equations::AbstractEquations{2}, basis::LobattoLegendreBasis, mesh::TreeMesh{2})
+
+  alpha = Vector{real(basis)}()
+  alpha_tmp = similar(alpha)
+
+  A = Array{real(basis), ndims(equations)}
+  indicator_threaded  = [A(undef, nnodes(basis), nnodes(basis)) for _ in 1:Threads.nthreads()]
+  modal_threaded      = [A(undef, nnodes(basis), nnodes(basis)) for _ in 1:Threads.nthreads()]
+  modal_tmp1_threaded = [A(undef, nnodes(basis), nnodes(basis)) for _ in 1:Threads.nthreads()]
+
+  return (; alpha, alpha_tmp, indicator_threaded, modal_threaded, modal_tmp1_threaded, mesh)
+end
+
+# this method is used when the indicator is constructed as for AMR
+function create_cache(typ::Type{IndicatorNNRH}, mesh, equations::AbstractEquations{2}, dg::DGSEM, cache)
+  create_cache(typ, equations, dg.basis)
+end
+
+
+function (indicator_nnrh::IndicatorNNRH)(u::AbstractArray{<:Any,4},
+                                                   equations, dg::DGSEM, cache;
+                                                   kwargs...)
+  @unpack alpha_max, alpha_min, alpha_smooth, alpha_continuous, alpha_amr, variable, network = indicator_nnrh
+  @unpack alpha, alpha_tmp, indicator_threaded, modal_threaded, modal_tmp1_threaded, mesh = indicator_nnrh.cache
+  # TODO: Taal refactor, when to `resize!` stuff changed possibly by AMR?
+  #       Shall we implement `resize!(semi::AbstractSemidiscretization, new_size)`
+  #       or just `resize!` whenever we call the relevant methods as we do now?
+  resize!(alpha, nelements(dg, cache))
+  if alpha_smooth
+    resize!(alpha_tmp, nelements(dg, cache))
+  end
+
+  c2e = zeros(Int, length(mesh.tree))
+  for element in eachelement(dg, cache)
+    c2e[cache.elements.cell_ids[element]] = element
+  end
+
+  network_input = Array{Float64}(undef, 3, nelements(dg, cache))
```

Comment:
This reallocates a large array each time the indicator is used, which is probably *very* expensive. Can you please check the timer and see if the indicator is indeed expensive and has a lot of allocations? If yes, we should get rid of this by preallocating it.